design/restic-backup-and-restore-progress.md

```diff
@@ -0,0 +1,104 @@
+The Goroutine will periodically read the stream (every 10 seconds) and get the last printed status line.
+The line will be decoded as JSON and the `percent_done` property will be read and formatted as a percentage value.
+
+The PodVolumeBackupStatus type will be extended to include a `Progress` field.
+The PodVolumeBackup will be patched to set the `status.Progress` field to the percentage value of the progress.
+
+Once the backup has completed successfully, the PodVolumeBackup will be patched to set `status.Progress = 100%`.
+
+### restic restore progress
+
+The `restic ls <snapshot_id> --json` command provides information about the size of backed up files:
+
+```
+{"time":"2019-08-14T15:16:29.105557-07:00","tree":"1de5381473aecb5548959f6e5444bb7966fe781500b5f8ec1629088674bea6a6","paths":["/Users/aadnan/Playground/restic/files"],"hostname":"aadnan-a01.vmware.com","username":"aadnan","uid":501,"gid":20,"id":"5628926b4f0e105eb33d83c619f235a8cf8f695d19955fab44bf506ece195f3a","short_id":"5628926b","struct_type":"snapshot"}
+{"name":"files","type":"dir","path":"/files","uid":501,"gid":20,"mode":2147484096,"mtime":"2019-08-14T15:16:20.53091354-07:00","atime":"2019-08-14T15:16:20.53091354-07:00","ctime":"2019-08-14T15:16:20.53091354-07:00","struct_type":"node"}
+{"name":"file1","type":"file","path":"/files/file1","uid":501,"gid":20,"size":1435500544,"mode":384,"mtime":"2019-08-14T15:16:20.491206506-07:00","atime":"2019-08-14T15:16:20.491206506-07:00","ctime":"2019-08-14T15:16:20.491234451-07:00","struct_type":"node"}
+...
+```
+
+Before beginning the restore operation, we can use the output of `restic ls` to get a list of the files that will be restored and calculate the total size of the backup.
+
+The code to run the `restic restore` command will be changed to include a Goroutine that periodically (every 10 seconds) goes through the list of files in the backup and gets the current size of each file in the volume.
+If the file doesn't exist in the volume yet (restic hasn't started restoring it), the size is 0.
+The sum of the current size of each file is then compared with the total size of the backup to calculate a completion percentage (`cursize / totalsize * 100`).
+
+The PodVolumeRestoreStatus type will be extended to include a `Progress` field.
+The PodVolumeRestore will be patched to set the `status.Progress` field to the percentage value of the progress.
+
+Once the restore has completed successfully, the PodVolumeRestore will be patched to set `status.Progress = 100%`.
+
+## Open Questions
+
+- Can we assume that the volume we are restoring in will be empty? Can it contain other artefacts? If we can assume it to be empty, we can go with the easier approach detailed below.
+
+## Alternatives Considered
+
+### restic restore progress
+
+If we can assume that the volume we are restoring into will be empty, we can instead use the output from `restic stats` to get the total size of the backup.
```

Comment:
I do like the simplicity of this approach, since we don't have to do any kind of per-file processing (which is more complicated and potentially has performance issues for large volumes).  We can use something like `du -s` to get the total current size of the volume to compare to the `restic stats` total size.

I *believe* we can assume that the volume is empty when a restic restore starts - in the case of PVCs, we dynamically provision an empty volume, and for things like `emptyDir` we'd be getting an empty volume as well. Need to think about this a little more, though.

Reply:
Actually I'm not sure that `du -s` can be used here, when I tested this locally the size reported from `du` is a little different due to it showing actual disk usage. There is `du --apparent-size` which is supposed to give the "apparent" size of the directory, but even with that flag I'm still seeing a disparity.

I did find https://github.com/missedone/dugo/blob/master/du.go which walks the directory and gets the total size of each file, this does end up matching what `restic stats --json` reports. However, if we're having to walk the directory anyway, we might as well only walk the files listed in the backup.

I'll look more into `du` though to see if I can figure out how to get it to match the actual size, because it would certainly be more performant for large volumes.

Reply:
Ah, OK. It would be nice not to have to check every file every time - not sure what the performance would be like for volumes with large #'s of files.

Reply:
Okay so the discrepancy between `du --apparent-size --block-size 1` and restic's size comes from `du` counting the size of the directory entry as well as the files.

For example:
```
$ ls -la
total 11277312
drwx------  7 aadnan  staff         224 Aug 19 15:50 .
drwxr-xr-x  6 aadnan  staff         192 Aug 14 14:47 ..
-rw-------  1 aadnan  staff  1435500544 Aug 14 15:16 file1
-rw-------  1 aadnan  staff  1317011456 Aug 14 15:16 file2
-rw-------  1 aadnan  staff  1115684864 Aug 14 15:16 file3
-rw-------  1 aadnan  staff  1340080128 Aug 14 15:16 file4
-rw-------  1 aadnan  staff   523239424 Aug 14 15:16 file5
```

For the above directory, the restic backup size would be `5731516416` which is the sum of the sizes for file1-5. The calculation from `du -b` (equivalent to `du --apparent-size --block-size 1`) produces `5731516640` which is the sum of the sizes for file1-5 plus the `224`, the size of the current directory entry.

The problem is more difficult for subdirectories as the total sum will include the size of each nested sub-directory entry and there doesn't seem to be a way to get `du` to only consider files. The only way I can find to do this is to use `find` to pass the list of files in the directory to `du` and use the last line (https://superuser.com/questions/853667/getting-du-to-ignore-size-of-and-entries):

```
$ du -c -b `find . -type f`
1115684864	./file3
1340080128	./file4
522190848	./test/file2
1326448640	./test/subdir/file2
1810890752	./test/subdir/file1
1167065088	./test/file1
523239424	./file5
1317011456	./file2
1435500544	./file1
10558111744	total
```

This total matches the restic total size for the backup:
```
$ restic stats --json latest
{"total_size":10558111744,"total_file_count":11}
```

I'm not sure how performant this will be compared to walking the tree in Go directly, especially for larger volumes. I also don't know that `du` has any special implementation that allows it to not have to walk through each file to calculate the total size (looks like it's doing just this from the implementation: http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/du.c#n678). Looks likes `du` can take a long time on large volumes too (https://unix.stackexchange.com/questions/8485/how-to-cache-or-otherwise-speed-up-du-summaries - > 2 minutes in this case). However, it does sound like repeated calls to `du` are faster due to the filesystem buffer - I imagine this would apply to Go's direct access as well since this isn't specific to `du`.

The effect of this in larger volumes is that it will take longer to update the progress, but the restore itself will also take longer in these cases, so this might be okay?

Reply:
OK, so sounds like there's probably not really benefit in using `du` over walking the file tree in golang.  I'm fine proceeding with the latter for now, but we'll definitely want to find a way to do some benchmarking and see what the performance is like. We could possibly use `inotify` to get notifications of file changes rather than having to scan, but that's likely much more complex implementation-wise so I wouldn't want to jump there unless needed (and more  research done).

Reply:
It also seems to me like using `restic stats` to get the total bytes to be restored should be fine. I don't think there's a scenario where velero would restore into a non-empty directory (we dynamically re-provision new PVs to restore into -- https://github.com/heptio/velero/blob/master/pkg/restore/restore.go#L896). Even if it contains a small number of non-empty files, our "bytes done" estimate would just be slightly inflated, probably not the end of the world. Does that seem reasonable?

Reply:
Yep, that all sounds good to me. Not a real benchmark, but with a quick try in a 310GB directory on my local machine `du` and `dugo` appear to have similar execution times (after purging disk buffers):

```
$ sudo purge
$ time du -sh
310G	.
du -sh  0.00s user 0.01s system 54% cpu 0.014 total
$ sudo purge
$ time dugo -h .
310.0G	 ./
dugo -h .  0.00s user 0.00s system 48% cpu 0.014 total
```

> We could possibly use inotify to get notifications of file changes rather than having to scan

That's definitely an interesting idea, but agree that we don't really need to consider that unless we see performance issues in the current approach.

> Even if it contains a small number of non-empty files, our "bytes done" estimate would just be slightly inflated, probably not the end of the world. Does that seem reasonable?

That sounds good to me, given we'll set BytesDone = TotalBytes once the restic operation completes successfully, any residual bytes probably won't even be exposed to the user.

Reply:
Okay updated the proposal to reflect what we've discussed here.